---
title: "Querying iAtlas Data"
output: html_notebook
---

The data hosted by iAtlas is available through an API.

```{r}
# git clone the notebook repo to get this file #
source('https://raw.githubusercontent.com/CRI-iAtlas/iatlas-notebooks/general_db_query/functions/notebook_functions.R')
#source('functions/notebook_functions.R')
library_setup()
```

Exploring the datasets and features

The iAtlas data is stored in a database that can be queried with functions from the
iatlasGraphQLClient package. We have clinical data, immune features, scores of
predictors of response to immunotherapy, and quantile normalized gene expression.
You can get more information in iAtlas on immune features, and our annotation of
immunomodulators genes. You can access more information about these
datasets in iAtlas.

As a first step, letâ€™s take a look at the available datasets.

```{r}
# datasets that we have in the iAtlas database
datasets <- iatlasGraphQLClient::query_datasets()
datasets
```

Let's count the datasets by type.


```{r}
table(datasets$type)
```

Currently there are three types: analysis, ici, and other. Analysis data sets include TCGA and PCAWG data while ici
datasets relate to immune checkpoint inhibitor studies. Currently, there is only one 'other' type:
the Genotype-Tissue Expression (GTEx) project which is a resource aimed at healthy normal tissue expression.

Data can be tagged with different descriptive terms which help in finding appropriate data given a particular interest.

```{r}
# datasets that we have in the iAtlas database
dataset_tags <- iatlasGraphQLClient::query_dataset_tags()

# count up the tags
table(dataset_tags$tag_name)
```

Suppose, we're interested in data sets that are tagged as "response data".

```{r}
dataset_tags[dataset_tags$tag_name == 'TCGA_Study',]
```

From that list, we might be interested in a particular dataset.
In order to get the sample IDs associated with that dataset, we use
the query_dataset_samples function.

```{r}
dataset_samples <- iatlasGraphQLClient::query_dataset_samples(datasets = "TCGA")
dim(dataset_samples)
```
With that table, we see that the dataset contains 17 samples.

In order to learn something about those samples, we'll query the sample metadata.

```{r}
# running this function with no parameters will return the entire table.
sample_metatdata <- iatlasGraphQLClient::query_sample_patients(samples = dataset_samples$sample_name) #

dim(sample_metatdata)
```
In this case we can get patient sex and race.

How many samples per patient?

```{r}
sample_metatdata %>% select(patient_name, sample_name) %>% unique() %>% dim()
```

OK, so there's one sample per patient.


Get immunomodulator genes.
query_immunomodulators
query_genes
query_io_targets

Use those to query whether these samples had an mutations in those genes.
query_mutation_statuses

Pick a gene where there was a mutation present,
then pull down expression data for that gene and visualize it.


OK, so, there's 11,080 samples in the TCGA dataset. Let's see the top of the dataset_samples table.

```{r}
head(dataset_samples)
```


```{r}
d = iatlasGraphQLClient::query_cohorts()
d[d$dataset_name == 'TCGA',]
```
```{r}
tcga_studies <- iatlasGraphQLClient::query_tags_with_parent_tags(parent_tags = "TCGA_Study")
tcga_studies
```
```{r}
d = iatlasGraphQLClient::query_tag_samples(parent_tags = "TCGA_Study",tags= c(tag_name = "KICH"))
d
```
